Add Cuda half_t type and operators

e10harvey opened the pull request.

crtrott commented:
Evan reports this issue with the bool conversion in there: 
```
Kokkos_Cuda_Half.hpp(109): error: more than one instance of constructor "__half::__half" matches the argument list:
            function "__half::__half(int)"
            function "__half::__half(__half &&)"
            argument types are: (Kokkos::Experimental::half_t)
          detected during instantiation of "Kokkos::Experimental::half_t::half_t(T) [with T=int]" 
```

crtrott commented:
The general goal is that half_t will behave just like double or float with respect to implicit conversions also with integer and bool.

dalg24 commented on core/src/Cuda/Kokkos_Cuda_Half.hpp:
```diff
@@ -49,14 +49,211 @@
+  __half val;
+
+public:
+  KOKKOS_FORCEINLINE_FUNCTION
```
```suggestion
```
Functions are defined within the body of the class which is implicitly `inline`.  Do you good reason for the "force"?

e10harvey replied:
No. I also think the force can be removed.

dalg24 commented on core/src/Cuda/Kokkos_Cuda_Half.hpp:
```diff
@@ -49,14 +49,211 @@
+
+  // Binary operators
+  KOKKOS_INLINE_FUNCTION
+  half_t operator= (__half rhs) {
```
Shouldn't it be returning a reference?

e10harvey replied:
I don't think so since this is a scalar type. Thoughts?

crtrott replied:
This needs to return a reference. Consider this code:
```
#include<cstdio>
#include<cstdlib>
int main(int argc, char* argv[]) {
  double x = 1.0, y=2.0;
  double* ptr = &(x=y);
  printf("%p %p %lf\n",ptr,&x,x);
}
```
Prints:
```
0x7fffe4133b18 0x7fffe4133b18 2.000000
```

dalg24 commented on core/src/Cuda/Kokkos_Cuda_Half.hpp:
```diff
@@ -49,14 +49,211 @@
+
+  // Logical operators
+  KOKKOS_INLINE_FUNCTION
+  half_t operator! () const {
```
Do you actually need that one?

e10harvey replied:
I think so, !a and !!a for `float a` result in 0x0 regardless of what value a is.

dalg24 commented on core/unit_test/TestHalfOperators.hpp:
```diff
@@ -0,0 +1,186 @@
+    d_rhs = Kokkos::Experimental::cast_from_half<double>(rhs);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const int &i) const {
```
```suggestion
  KOKKOS_FUNCTION
  void operator()(int) const {
```

dalg24 commented on core/unit_test/TestHalfOperators.hpp:
```diff
@@ -0,0 +1,186 @@
+#define FP16_EPSILON 0.0009765625F
+#define FP16_MAX 65504.0F           // Maximum normalized number
+#define FP16_MIN 0.000000059604645F // Minimum normalized positive half precision number
+using half_t = Kokkos::Experimental::half_t;
```
```suggestion
using Kokkos::Experimental::half_t;
```

dalg24 commented on core/unit_test/TestHalfOperators.hpp:
```diff
@@ -0,0 +1,186 @@
+  double d_lhs, d_rhs;
+  ViewType actual_lhs, expected_lhs;
+
+  Functor_TestHalfOperators():
+    lhs(0),
+    rhs(0)
+  {
+    actual_lhs = ViewType("actual_lhs", N_OPS);
+    expected_lhs = ViewType("expected_lhs", N_OPS);
+    d_lhs = Kokkos::Experimental::cast_from_half<double>(lhs);
+    d_rhs = Kokkos::Experimental::cast_from_half<double>(rhs);
+  }
+
+  Functor_TestHalfOperators(half_t lhs, half_t rhs):
+    lhs(lhs),
+    rhs(rhs)
+  {
+    actual_lhs = ViewType("actual_lhs", N_OPS);
+    expected_lhs = ViewType("expected_lhs", N_OPS);
+    d_lhs = Kokkos::Experimental::cast_from_half<double>(lhs);
+    d_rhs = Kokkos::Experimental::cast_from_half<double>(rhs);
+  }
```
```suggestion
  Functor_TestHalfOperators(half_t lhs = 0, half_t rhs = 0):
    lhs(lhs),
    rhs(rhs)
  {
    actual_lhs = ViewType("actual_lhs", N_OPS);
    expected_lhs = ViewType("expected_lhs", N_OPS);
    d_lhs = Kokkos::Experimental::cast_from_half<double>(lhs);
    d_rhs = Kokkos::Experimental::cast_from_half<double>(rhs);
  }
```

dalg24 commented on core/unit_test/TestHalfOperators.hpp:
```diff
@@ -0,0 +1,186 @@
+    actual_lhs(ADD) = Kokkos::Experimental::cast_from_half<double>(lhs + rhs);
+    expected_lhs(ADD) = d_lhs + d_rhs;
+
+    actual_lhs(ADD) = Kokkos::Experimental::cast_from_half<double>(lhs + rhs);
+    expected_lhs(ADD) = d_lhs + d_rhs;
+
```
```suggestion
```

e10harvey replied:
@dalg24: no need to test?

dalg24 replied:
This duplicated L96-97 and L99-100

dalg24 commented on core/unit_test/TestHalfOperators.hpp:
```diff
@@ -0,0 +1,186 @@
+
+  inline
+  void run_on_device() {
+    Kokkos::parallel_for(
+            "Test::Functor_TestHalfOperators",
+            Kokkos::RangePolicy<>(0,1),
+            *this);
```
I would stick this directly in the constructor.

dalg24 replied:
Also we might want to be explicit about the execution space.

crtrott replied:
I was only sticking this into the default test executable which uses default execution space.

dalg24 reviewed (commented):
Probably should implement operator== too

crtrott commented on core/src/Cuda/Kokkos_Cuda_Half.hpp:
```diff
@@ -49,14 +49,270 @@
+  // Cast rhs to half for assignment to lhs of type half_t
+  template <class T>
+  KOKKOS_FUNCTION
+  half_t(T rhs):val(cast_to_half(rhs)) {}
```
the problem with the above bool conversion put in goes away if we do constructor forwarding here:
```
half_t(T rhs):half_t(cast_to_half(rhs)) {}
```

crtrott commented on core/src/Cuda/Kokkos_Cuda_Half.hpp:
```diff
@@ -49,14 +49,270 @@
+  }
+
+  KOKKOS_FUNCTION
+  half_t operator+ (__half rhs) const {
```
all of those should take half_t not __half and then you need to use rhs.val below.

crtrott reviewed (changes requested):
the operators should all take half_t not __half (saves on conversion operators) and we need the compound ones += -= *= /=.

e10harvey committed:
Update Half Precision support for CUDA

Adds operator overloads for half_t.

Adds support for half_t arithmetic on host via promotion to float.

e10harvey committed:
Adds Half Precision operator tests
